File: src/main/java/me/st28/flexseries/flexlib/command/CommandModule.java
/**
 * Copyright 2016 Stealth2800 <http://stealthyone.com/>
 * Copyright 2016 Contributors <https://github.com/FlexSeries>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.st28.flexseries.flexlib.command;

import me.st28.flexseries.flexlib.FlexLib;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.BooleanResolver;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.DoubleResolver;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.FloatResolver;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.IntegerResolver;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.LongResolver;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.PlayerResolver;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.SessionResolver;
import me.st28.flexseries.flexlib.command.DefaultArgumentResolvers.StringResolver;
import me.st28.flexseries.flexlib.command.argument.ArgumentResolver;
import me.st28.flexseries.flexlib.plugin.FlexModule;
import me.st28.flexseries.flexlib.plugin.FlexPlugin;
import org.apache.commons.lang.Validate;
import org.bukkit.command.CommandSender;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public final class CommandModule extends FlexModule<FlexLib> {

    // Welcome to nested map hell, enjoy your stay

    final Map<Class<? extends FlexPlugin>, Map<String, FlexCommand>> commands = new HashMap<>();

    // < Plugin name + ID, < CommandSender, Session > >
    final Map<String, Map<String, CommandSession>> sessions = new HashMap<>();

    public CommandModule(FlexLib plugin) {
        super(plugin, "commands", "Manages the FlexLib command framework");
    }

    @Override
    protected void handleEnable() {
        ArgumentResolver.register(null, "boolean", new BooleanResolver());
        ArgumentResolver.register(null, "integer", new IntegerResolver());
        ArgumentResolver.register(null, "long", new LongResolver());
        ArgumentResolver.register(null, "float", new FloatResolver());
        ArgumentResolver.register(null, "double", new DoubleResolver());
        ArgumentResolver.register(null, "player", new PlayerResolver());
        ArgumentResolver.register(null, "string", new StringResolver());
        ArgumentResolver.register(null, "session", new SessionResolver());
    }

    public BasicCommand getCommand(Class<? extends FlexPlugin> plugin, String command) {
        Validate.notNull(plugin, "Plugin cannot be null");
        Validate.notNull(command, "Command cannot be null");

        if (!commands.containsKey(plugin)) {
            return null;
        }

        final String[] path = command.split(" ");

        BasicCommand found = commands.get(plugin).get(path[0].toLowerCase());
        for (int i = 1; i < path.length; ++i) {
            found = found.subcommands.get(path[i].toLowerCase());
            if (found == null) {
                break;
            }
        }
        return found;
    }

    void registerCommand(Class<? extends FlexPlugin> plugin, FlexCommand command) {
        if (!commands.containsKey(plugin)) {
            commands.put(plugin, new HashMap<>());
        }

        commands.get(plugin).put(command.label.toLowerCase(), command);
    }

    /**
     * Retrieves a specified {@link CommandSession}.
     * @param plugin The plugin the session/command belongs to.
     * @param sender The CommandSender to retrieve the session of.
     * @param id The ID of the session.
     * @param create True to create a new session object if it doesn't exist.
     * @return A {@link CommandSession} matching the given parameters.
     *         Null if the session doesn't exist and <code>create</code> is false
     */
    public CommandSession getSession(Class<? extends FlexPlugin> plugin, CommandSender sender, String id, boolean create) {
        final String key = plugin.getCanonicalName() + "#" + id;
        final String subkey = sender.getClass().getCanonicalName() + "#" + sender.getName();

        if (!sessions.containsKey(key)) {
            sessions.put(key, new HashMap<>());
        }

        final Map<String, CommandSession> submap = sessions.get(key);
        if (!submap.containsKey(subkey) && create) {
            submap.put(subkey, new CommandSession());
        }
        return submap.get(subkey);
    }

}

File: src/main/kotlin/me/st28/flexseries/flexlib/command/CommandModule.kt
/**
 * Copyright 2016 Stealth2800 <http://stealthyone.com/>
 * Copyright 2016 Contributors <https://github.com/FlexSeries>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.st28.flexseries.flexlib.command

import me.st28.flexseries.flexlib.FlexLib
import me.st28.flexseries.flexlib.command.argument.*
import me.st28.flexseries.flexlib.logging.LogHelper
import me.st28.flexseries.flexlib.plugin.FlexModule
import me.st28.flexseries.flexlib.plugin.FlexPlugin
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerQuitEvent
import java.math.BigDecimal
import java.util.*
import kotlin.reflect.KClass
import kotlin.reflect.KType
import kotlin.reflect.defaultType

/***
 * The main command manager for the FlexLib command framework.
 */
class CommandModule(plugin: FlexLib) : FlexModule<FlexLib>(plugin, "commands", "Manages the FlexLib command framework"), Listener {

    internal val commands: MutableMap<KClass<out FlexPlugin>, MutableMap<String, FlexCommand>> = HashMap()

    internal val argumentParsers: MutableMap<String, ArgumentParser<Any>> = HashMap()

    // Sender identifier, < command UUID -> session >
    internal val sessions: MutableMap<String, MutableMap<UUID, CommandSession>> = HashMap()

    override fun handleEnable() {
        registerArgumentParser(Int::class, IntParser)
        registerArgumentParser(Double::class, DoubleParser)
        registerArgumentParser(BigDecimal::class, BigDecimalParser)
        registerArgumentParser(Player::class, PlayerParser)
        registerArgumentParser(String::class, StringParser)
    }

    fun getBaseCommand(plugin: KClass<out FlexPlugin>, command: String): FlexCommand? {
        if (!commands.containsKey(plugin)) {
            return null
        }
        return commands[plugin]!![command]
    }

    fun registerCommand(command: FlexCommand) {
        val plugin = command.plugin.javaClass.kotlin
        if (!commands.containsKey(plugin)) {
            commands.put(plugin, HashMap())
        }

        commands[plugin]!!.put(command.label.toLowerCase(), command)
    }

    /**
     * Registers an [ArgumentParser].
     *
     * @param type The class the parser handles.
     * @param parser The parser for the specified class type.
     * @return True if the parser was successfully registered.
     *         False if there already is a parser registered for the specified type.
     */
    fun <T: Any> registerArgumentParser(type: KClass<T>, parser: ArgumentParser<T>): Boolean {
        val key = type.defaultType.toString()
        if (argumentParsers.containsKey(key)) {
            return false
        }

        argumentParsers.put(key, parser)
        return true
    }

    fun getArgumentParser(type: KType): ArgumentParser<Any>? {
        return argumentParsers[type.toString().replace("?", "")]
    }

    /**
     * Creates a session for the specified command and sender.
     *
     * @return The newly created session.
     *         Null if a session already exists.
     */
    internal fun createSession(command: BasicCommand, context: CommandContext): CommandSession? {
        val identifier = (context.sender as? Player)?.uniqueId?.toString() ?: "CONSOLE"
        val pSessions = sessions.getOrPut(identifier, { HashMap() })

        if (pSessions[command.uuid]?.running ?: false) {
            return null
        }

        pSessions.put(command.uuid, CommandSession(plugin, context))
        return pSessions[command.uuid]!!
    }

    @EventHandler
    fun onPlayerQuit(e: PlayerQuitEvent) {
        val sessions = this.sessions.remove(e.player.uniqueId.toString()) ?: return
        if (sessions.isEmpty()) {
            return
        }

        LogHelper.debug(this, "Cancelling command sessions for ${e.player.name}")
        for (v in sessions.values) {
            v.running = false
        }
    }

}